okhttp/src/main/java/okhttp3/internal/ws/MessageDeflater.kt

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.ws
+
+import java.io.Closeable
+import java.io.IOException
+import java.util.zip.Deflater
+import okio.Buffer
+import okio.ByteString
+import okio.ByteString.Companion.decodeHex
+import okio.DeflaterSink
+
+private val EMPTY_DEFLATE_BLOCK = "000000ffff".decodeHex()
+private const val LAST_OCTETS_COUNT_TO_REMOVE_AFTER_DEFLATION = 4
+
+class MessageDeflater(
```

Comment:
Nothing about these types are specific to WS or its messages. Should this be a BufferDeflater?

Reply:
It manipulates the deflated data to match the WS spec and is unlikely to be appropriate in other contexts. I don’t think it’s general purpose!